include/libtorrent/aux_/win_cng.hpp

```diff
@@ -0,0 +1,206 @@
+/*
+
+Copyright (c) 2016, Arvid Norberg
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_WIN_CNG_HPP
+#define TORRENT_WIN_CNG_HPP
+
+#include <vector>
+
+#include "libtorrent/config.hpp"
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/aux_/throw.hpp"
+#include "libtorrent/aux_/windows.hpp"
+
+#include "libtorrent/aux_/disable_warnings_push.hpp"
+#include <bcrypt.h>
+#include "libtorrent/aux_/disable_warnings_pop.hpp"
+
+namespace libtorrent { namespace aux {
+
+	static void throw_ntstatus_error(char const* name, NTSTATUS status) {
+		char message[1024];
+		std::snprintf(message, sizeof(message),
+			"Error 0x%x returned by %s", status, name);
+		throw_ex<std::runtime_error>(message);
```

Comment:
instead of converting the `NTSTATUS` error message into a string, isn't there an `error_category` for NT error codes? Ideally it would be preserved in the error message (and ideally the string would only be generated on demand, when `what()` is called for instance)

Reply:
I thought NTSTATUS is not compatible with generic_category or system_category(windows getlasterror results), but have found MS related source that uses ntstatus as system_error code https://github.com/Azure/azure-storage-fuse/blob/43e82df5d85a4c082dc67af8131bcf05f4d9270a/azure-storage-cpp-lite/src/hash.cpp#L16

Will fix this

Reply:
great! then you could use the `system_error` exception too. unfortunately I don't think you can use `std::error_code` as `boost::system::error_code` is used pervasively throughout boost.asio